runtime/src/bank.rs

```diff
@@ -8131,26 +8170,79 @@ impl Bank {
             if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
                 datapoint_info!(datapoint_name, ("slot", self.slot, i64));
 
-                // Burn lamports in the old account
-                self.capitalization
-                    .fetch_sub(old_account.lamports(), Relaxed);
-
-                // Transfer new account to old account
-                self.store_account(old_address, &new_account);
-
-                // Clear new account
-                self.store_account(new_address, &AccountSharedData::default());
+                self.replace_account(old_address, new_address, Some(&old_account), &new_account);
 
                 // Unload a program from the bank's cache
                 self.loaded_programs_cache
                     .write()
                     .unwrap()
                     .remove_programs([*old_address].into_iter());
+            }
+        }
+    }
 
-                self.calculate_and_update_accounts_data_size_delta_off_chain(
-                    old_account.data().len(),
-                    new_account.data().len(),
-                );
+    /// Use to replace an empty account with a program by feature activation
+    fn replace_empty_account_with_upgradeable_program(
+        &mut self,
+        old_address: &Pubkey,
+        new_address: &Pubkey,
+        datapoint_name: &'static str,
+    ) {
+        // Must be attempting to replace an empty account with a program
+        // account _and_ data account
+        if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
+            let (old_data_address, _) = Pubkey::find_program_address(
+                &[old_address.as_ref()],
+                &bpf_loader_upgradeable::id(),
+            );
+            let (new_data_address, _) = Pubkey::find_program_address(
+                &[new_address.as_ref()],
+                &bpf_loader_upgradeable::id(),
+            );
+            if let Some(new_data_account) = self.get_account_with_fixed_root(&new_data_address) {
+                // Make sure the old account is empty
+                // We aren't going to check that there isn't a data account at
+                // the known program-derived address (ie. `old_data_address`),
+                // because if it exists, it will be overwritten
+                if self.get_account_with_fixed_root(old_address).is_none() {
+                    let data = old_data_address.as_ref().to_vec();
+                    let lamports = self.get_minimum_balance_for_rent_exemption(data.len());
+
+                    // Make sure the new account has enough rent-exempt
+                    // lamports for the empty account that will now house the
+                    // PDA
+                    if new_account.lamports() >= lamports {
+                        if let Some(change_in_cap) = new_account.lamports().checked_sub(lamports) {
```

Comment:
You can just use a `saturating_sub` here since you already checked that `new_account.lamports() >= lamports`